renderdoc/driver/ihv/intel/intel_counters.cpp

```diff
@@ -0,0 +1,423 @@
+
+  HRESULT hr = m_device->CreateCounter(&counter_desc, &m_counter);
+  if(FAILED(hr))
+  {
+    m_subscribedMetricSets[m_passIndex]->Deactivate();
+    return;
+  }
+
+  res = m_subscribedMetricSets[m_passIndex]->Deactivate();
+  if(res != TCompletionCode::CC_OK)
+    return;
+
+  m_deviceContext->Begin(m_counter);
+}
+
+void IntelCounters::EndSample()
+{
+  if(!m_metricsDevice)
+    return;
+
+  m_deviceContext->End(m_counter);
+  HRESULT hr = S_OK;
+  uint32_t iteration = 0;
+  const uint32_t max_attempts = 0xFFFFFFFF;
+  void *counter_data = nullptr;
+
+  do
+  {
+    uint32_t flags = (iteration == 0) ? 0 : D3D11_ASYNC_GETDATA_DONOTFLUSH;
+    hr = m_deviceContext->GetData(m_counter, &counter_data, m_counter->GetDataSize(), flags);
+    iteration++;
+  } while(hr != S_OK && iteration < max_attempts);
+  m_counter->Release();
+
+  if(hr != S_OK)
+    return;
+
+  IMetricSet_1_1 *metricSet = m_subscribedMetricSets[m_passIndex];
+  TMetricSetParams_1_0 *metricSetParams = metricSet->GetParams();
+
+  uint32_t calculatedReportCount = 0;
+  TCompletionCode res = m_subscribedMetricSets[m_passIndex]->CalculateMetrics(
+      (const unsigned char *)counter_data, metricSetParams->QueryReportSize, m_queryResult.data(),
+      (uint32_t)m_queryResult.size() * sizeof(TTypedValue_1_0), &calculatedReportCount, false);
+
+  if(res != TCompletionCode::CC_OK)
+    return;
+
+  for(int i = 0; i < m_subscribedMetricsByCounterSet[m_passIndex].size(); i++)
+  {
+    GPUCounter counterId = m_subscribedMetricsByCounterSet[m_passIndex][i];
+    uint32_t counterIndex = m_metricLocation[counterId].second;
+
+    m_results[std::pair<GPUCounter, uint32_t>(counterId, m_sampleIndex)] =
+        m_queryResult[counterIndex];
+  }
+  m_sampleIndex++;
+}
+
+std::vector<CounterResult> IntelCounters::GetCounterData(const std::vector<uint32_t> &eventIDs,
+                                                         const std::vector<GPUCounter> &counters)
+{
+  std::vector<CounterResult> ret;
+  for(uint32_t sample = 0; sample < (uint32_t)eventIDs.size(); sample++)
+  {
+    for(size_t counter = 0; counter < counters.size(); counter++)
+    {
+      const CounterDescription desc = GetCounterDescription(counters[counter]);
+
+      switch(desc.resultType)
+      {
+        case CompType::UInt:
+        {
+          if(desc.resultByteWidth == sizeof(uint32_t))
+          {
+            uint32_t value =
+                m_results[std::pair<GPUCounter, uint32_t>(desc.counter, sample)].ValueUInt32;
+            if(desc.unit == CounterUnit::Percentage)
+            {
+              value = RDCCLAMP(value, 0U, 100U);
+            }
+
+            ret.push_back(CounterResult(eventIDs[sample], counters[counter], value));
+          }
+          else if(desc.resultByteWidth == sizeof(uint64_t))
+          {
+            uint64_t value =
+                m_results[std::pair<GPUCounter, uint32_t>(desc.counter, sample)].ValueUInt64;
+            // RenderDoc expects time metrics in microceconds
```

Comment:
This comment is odd. For seconds units the values are expected to come back in seconds normalised (same as for bytes etc). If the counters are in nanoseconds as the other code implies this should divide by 1,000,000,000. You'd also want to convert to double first so you don't lose precision.